### PR TITLE
DOC: Adding script to validate docstrings, and generate list of all functions/methods with state

### PR DESCRIPTION
**Pull request by datapythonista**
- [ ] closes #xxxx
- [ ] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Adding a new script to check whether a docstring follows the defined standards or not. Also, without arguments, the script returns a report of all the functions and methods in the API, with information about its docstring.

CC @jorisvandenbossche

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
Cool! 

Does the running of the doctests already work? (I don't seem to get output of that)

**Comment by jorisvandenbossche** [comment]
> Does the running of the doctests already work? (I don't seem to get output of that)

Ah, I see it is in the code, but not yet enable to check

**Comment by jorisvandenbossche** [comment]
Some quick feedback:

- Not all methods seems to be found, eg Series.mean is missing from the output
- It might be good to sort the methods/output (although of course when opening in a spreadsheet, it can easily be done afterwards)
- I think this can be a really useful overview to start from to make a todo list for the sprint

**Comment by TomAugspurger** [comment]
> because of something with indentation

presumably that would be fixed by setting the `NORMALIZE_WHITESPACE` flag?

FYI, I had to make the file executable with `chmod +x` so that that `./scripts/validate_docstrings.py` worked.

**Comment by TomAugspurger** [comment]
Specifically, we could add `doctest_optionflags= NORMALIZE_WHITESPACE` to our `setup.cfg`. I don't think that interferes with regular `pytest` flags when you aren't doing a doctest.

**Comment by datapythonista** [comment]
I'll take a look at the tests now. Regarding the missing methods, I was filtering (supposed to be) duplicates, to avoid having twice the stuff implemented in `generic.py`, but it seems functions using the same wrapper as all the stats were also filtered.

I added a column to detect potential code with the same docstring instead. And I made some changes to try to also have the information for properties and some other code not implemented as a "regular" Python function or method.

**Comment by datapythonista** [comment]
You can see the output here if you want to take a quick look before running it locally:
https://docs.google.com/spreadsheets/d/1XK25QrVJLUwVIScvRcmysiRe0QRMpCPz78kHsCeOSw4/edit?usp=sharing

**Comment by datapythonista** [comment]
Didn't know about the `NORMALIZE_WHITESPACE` flag, thanks for it @TomAugspurger. So far I added it to the script.

@jorisvandenbossche while I agree that having the input sorted would be nicer, I think it's simpler (and it doesn't need to load anything into memory) as it is now. The order is not random, is the one specified in `api.rst`, so it should make some sense. Also, sorting it is as simple as `./validate_docstrings.py | tail -n+2 | sort`. Unless you really think that sorting results is much better, I'll leave it like it is.

File: scripts/validate_docstrings.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+    @property
+    def deprecated(self):
+        pattern = re.compile('.. deprecated:: ')
+        return (self.method_name.startswith('pandas.Panel') or
+                bool(pattern.search(self.summary)) or
+                bool(pattern.search(self.extended_summary)))
+
+    @property
+    def examples_pass_tests(self):
+        flags = doctest.NORMALIZE_WHITESPACE
```
Comment (jorisvandenbossche) [inline scripts/validate_docstrings.py]: can you also add the `doctest.IGNORE_EXCEPTION_DETAIL` ? (I think it is summing them)

**Review by jorisvandenbossche — COMMENTED** [review]
Did play a bit with it, this will be really helpful I think. Added some further comments.

In the guidelines, I think we need to stress that this script is a guide, and in certain cases, for a good reason it is OK to not pass everything (like eg the case below).

One of the problems with the signature / keywords in docstring testing, is that there are some cases where this differs on purpose. Eg when using the `deprecate_kwarg` keyword, the deprecated one is not in the signature (it is automatically handled by the decorator), but often still listed as being deprecated.

File: scripts/validate_docstrings.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+        return not (bool(self.summary) and bool(self.extended_summary))
+
+    @property
+    def doc_parameters(self):
+        return self.doc['Parameters']
+
+    @property
+    def signature_parameters(self):
+        if not inspect.isfunction(self.method_obj):
+            return tuple()
```
Comment (jorisvandenbossche) [inline scripts/validate_docstrings.py]: Is there a reason you do this? As the code below works fine for a class as well, eg pandas.DataFrame (while now the script will say the keywords and signature did not match, while in fact they do)

Comment (datapythonista) [reply scripts/validate_docstrings.py]: I didn't research much, but if I introspect parameters from a non-function I get this error:

```
Traceback (most recent call last):
  File "./validate_docstrings.py", line 353, in <module>
    sys.exit(main(args.function))
  File "./validate_docstrings.py", line 337, in main
    return validate_all()
  File "./validate_docstrings.py", line 259, in validate_all
    int(doc.correct_parameters),
  File "./validate_docstrings.py", line 159, in correct_parameters
    return not bool(self.parameter_mismatches)
  File "./validate_docstrings.py", line 138, in parameter_mismatches
    signature_params = self.signature_parameters
  File "./validate_docstrings.py", line 130, in signature_parameters
    params = tuple(inspect.signature(self.method_obj).parameters.keys())
  File "/home/mgarcia/.anaconda3/envs/pandas_dev/lib/python3.6/inspect.py", line 3036, in signature
    return Signature.from_callable(obj, follow_wrapped=follow_wrapped)
  File "/home/mgarcia/.anaconda3/envs/pandas_dev/lib/python3.6/inspect.py", line 2786, in from_callable
    follow_wrapper_chains=follow_wrapped)
  File "/home/mgarcia/.anaconda3/envs/pandas_dev/lib/python3.6/inspect.py", line 2341, in _signature_from_callable
    'no signature found for builtin type {!r}'.format(obj))
ValueError: no signature found for builtin type <class 'pandas.core.indexing._AtIndexer'>
```

**Comment by datapythonista** [comment]
@jorisvandenbossche thanks for the review, useful comments. I think I addressed all them.

I think we'll want to keep adding validations as we find that people is making errors, but when you think this is in a good enough state, may be we can get it merged, so people working on the docstrings can start using it.

**Comment by jorisvandenbossche** [comment]
Great updates! Also good idea to print the docstring. Certainly for ones that are not composited of different parts / shared docstrings etc, it is good to see how the textual one is in totality. I added a commit to always try to dedent it.

> I think we'll want to keep adding validations as we find that people is making errors, but when you think this is in a good enough state, may be we can get it merged, so people working on the docstrings can start using it.

Indeed! Merging now.

**Comment by jorisvandenbossche** [comment]
Thanks a lot for this!

**Pull request merged**